cdp_client/cdp.py

```diff
@@ -48,10 +48,91 @@ class UnknownError(Exception):
     pass
 
 
+AuthResultCode = enum(
+    CREDENTIALS_REQUIRED=0,
+    # OK results:
+    GRANTED=1, # is also set when no authentication required
+    GRANTEDPASSWORDWILLEXPIRESOON = 2, # user should be notified about coming soon password expiry
+                                       # with suggestion to set a new password ASAP
+    # negative results:
+    NEWPASSWORDREQUIRED = 10, # password was OK but is expired, so new AuthRequest with additional
+                              # response with new password hash is required, and new password
+                              # complexity rules should be read from additionalCredentials[CredentialNewpassword].parameters
+    INVALIDCHALLENGERESPONSE = 11,
+    ADDITIONALRESPONSEREQUIRED = 12,
+    TEMPORARILYBLOCKED = 13,
+    REAUTHENTICATIONREQUIRED = 14 # server requires re-authentication (e.g.because of being idle),
+                                  # implementation should prompt the user for re-authentication
+                                  # (must not silently send cached credentials)
+)
+
+class UserAuthResult:
+    def __init__(self, code=AuthResultCode.CREDENTIALS_REQUIRED, text='', additional_credentials=None):
+        self._code = code
+        self._text = text
+        self._additional_credentials = additional_credentials
+
+    def code(self):
+        return self._code
+
+    def text(self):
+        return self._text
+
+    def additional_credentials(self):
+        return self._additional_credentials
+
+
+class AuthRequest(Promise):
+    def __init__(self, system_name='', application_name='', host='', port='', cdp_version='',
+                 system_use_notification=None, user_auth_result=UserAuthResult()):
+        self._system_name = system_name
+        self._application_name = application_name
+        self._host = host
+        self._port = port
+        self._cdp_version = cdp_version
+        self._system_use_notification = system_use_notification
+        self._user_auth_result = user_auth_result
+
+    def system_name(self):
+        return self._system_name
+
+    def application_name(self):
+        return self._application_name
+
+    def host(self):
+        return self._host
+
+    def port(self):
+        return self._port
+
+    def cdp_version(self):
+        return self._cdp_version
+
+    def user_auth_result(self):
+        return self._user_auth_result
+
+    def system_use_notification(self):
+        return self._system_use_notification
+
+    def accept(self, data=dict()):
+        self.do_resolve(data)
+
+    def reject(self):
+        self.do_reject(UnknownError('Authentication rejected'))
+
+
+class NotificationListener:
```

Comment:
handshake acceptance is not needed?

Reply:
Left out both notificationReceived() and handshakeAcceptanceRequested() (that exist in C++ client) form this change, as it would just complicate the client more and was currently able not to find any good use-case for them.
But they can be added later when we find some real use-case or need for them.